File: server/src/modules/llm/adapters/custom.adapter.ts
import { LLMAdapter, ChatMessage } from '../llm.interface';

export class CustomAdapter implements LLMAdapter {
  constructor(
    private apiKey: string,
    private baseUrl: string,
  ) {}

  async *chat(messages: ChatMessage[], stream: boolean): AsyncIterable<string> {
    // Custom adapter uses OpenAI-compatible API format
    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        messages,
        stream,
      }),
    });

    if (!stream) {
      const data = await response.json();
      yield data.choices?.[0]?.message?.content || data.content || '';
      return;
    }

    const reader = response.body?.getReader();
    if (!reader) return;

    const decoder = new TextDecoder();
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          const data = line.slice(6);
          if (data === '[DONE]') return;

          try {
            const parsed = JSON.parse(data);
            const content = parsed.choices?.[0]?.delta?.content;
            if (content) yield content;
          } catch {
            // Skip invalid JSON
          }
        }
      }
    }
  }
}

File: backend/services/llm_gateway/adapters/custom_adapter.py
"""
Custom API adapter for OpenAI-compatible endpoints.
"""
from typing import Iterator, List, Dict
import httpx
import json
from .base import BaseLLMAdapter


class CustomAdapter(BaseLLMAdapter):
    """Adapter for custom OpenAI-compatible APIs."""

    def __init__(self, api_key: str, model: str, base_url: str = None):
        super().__init__(api_key, model, base_url)
        if not base_url:
            raise ValueError("base_url is required for custom adapter")
        self.base_url = base_url.rstrip('/')

    def stream_chat(self, messages: List[Dict[str, str]]) -> Iterator[str]:
        """Stream chat completion from custom API."""
        url = f"{self.base_url}/chat/completions"
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
        payload = {
            "model": self.model,
            "messages": messages,
            "stream": True
        }

        with httpx.Client(timeout=60.0) as client:
            with client.stream("POST", url, headers=headers, json=payload) as response:
                response.raise_for_status()
                for line in response.iter_lines():
                    if line.startswith("data: "):
                        data = line[6:]
                        if data == "[DONE]":
                            break
                        try:
                            chunk = json.loads(data)
                            choices = chunk.get("choices", [])
                            if choices:
                                content = choices[0].get("delta", {}).get("content", "")
                                if content:
                                    yield content
                        except json.JSONDecodeError:
                            continue

    def chat(self, messages: List[Dict[str, str]]) -> str:
        """Get complete chat response from custom API."""
        url = f"{self.base_url}/chat/completions"
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
        payload = {
            "model": self.model,
            "messages": messages,
            "stream": False
        }

        with httpx.Client(timeout=60.0) as client:
            response = client.post(url, headers=headers, json=payload)
            response.raise_for_status()
            data = response.json()
            return data["choices"][0]["message"]["content"]

    def test_connection(self) -> str:
        """Test custom API connection."""
        self.chat([{"role": "user", "content": "Hello"}])
        return f"Connection successful. Model: {self.model}"
